src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -423,7 +424,10 @@ PeerImp::close()
 void
 PeerImp::fail(std::string const& reason)
 {
-    assert(strand_.running_in_this_thread());
+    if(! strand_.running_in_this_thread())
+        return strand_.post(std::bind (
+            (void(Peer::*)(std::string const&))&PeerImp::fail,
+                shared_from_this(), reason));
```

Comment:
Its got nothing to do with the behavior of `post` and everything to do with the behavior of `bind`. And `bind` always takes arguments by value.